File: Kernel Simulator/Misc/Screensaver/Displays/PersonLookup.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports KS.TimeDate
Imports Namer.NameGenerator

Namespace Misc.Screensaver.Displays
    Public Module PersonLookupSettings

        Private _personLookupDelay As Integer = 75
        Private _personLookupLookedUpDelay As Integer = 10000
        Private _personLookupMinimumNames As Integer = 10
        Private _personLookupMaximumNames As Integer = 100
        Private _personLookupMinimumAgeYears As Integer = 18
        Private _personLookupMaximumAgeYears As Integer = 100

        ''' <summary>
        ''' [PersonLookup] How many milliseconds to wait before getting the new name?
        ''' </summary>
        Public Property PersonLookupDelay As Integer
            Get
                Return _personLookupDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 75
                _personLookupDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [PersonLookup] How many milliseconds to show the looked up name?
        ''' </summary>
        Public Property PersonLookupLookedUpDelay As Integer
            Get
                Return _personLookupLookedUpDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 10000
                _personLookupLookedUpDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [PersonLookup] Minimum names count
        ''' </summary>
        Public Property PersonLookupMinimumNames As Integer
            Get
                Return _personLookupMinimumNames
            End Get
            Set(value As Integer)
                If value <= 10 Then value = 10
                If value > 1000 Then value = 1000
                _personLookupMinimumNames = value
            End Set
        End Property
        ''' <summary>
        ''' [PersonLookup] Maximum names count
        ''' </summary>
        Public Property PersonLookupMaximumNames As Integer
            Get
                Return _personLookupMaximumNames
            End Get
            Set(value As Integer)
                If value <= _personLookupMinimumNames Then value = _personLookupMinimumNames
                If value > 1000 Then value = 1000
                _personLookupMaximumNames = value
            End Set
        End Property
        ''' <summary>
        ''' [PersonLookup] Minimum age years
        ''' </summary>
        Public Property PersonLookupMinimumAgeYears As Integer
            Get
                Return _personLookupMinimumAgeYears
            End Get
            Set(value As Integer)
                If value <= 18 Then value = 18
                If value > 100 Then value = 100
                _personLookupMinimumAgeYears = value
            End Set
        End Property
        ''' <summary>
        ''' [PersonLookup] Maximum age years
        ''' </summary>
        Public Property PersonLookupMaximumAgeYears As Integer
            Get
                Return _personLookupMaximumAgeYears
            End Get
            Set(value As Integer)
                If value <= _personLookupMinimumAgeYears Then value = _personLookupMinimumAgeYears
                If value > 100 Then value = 100
                _personLookupMaximumAgeYears = value
            End Set
        End Property

    End Module

    Public Class PersonLookupDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Public Overrides Property ScreensaverName As String = "PersonLookup" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            PopulateNames()
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            Console.BackgroundColor = ConsoleColor.Black
            Console.ForegroundColor = ConsoleColor.Green
            Console.Clear()
            Console.CursorVisible = False

            'Generate names
            Dim NumberOfPeople As Integer = RandomDriver.Next(PersonLookupMinimumNames, PersonLookupMaximumNames)
            Dim NamesToLookup As String() = GenerateNames(NumberOfPeople)

            'Loop through names
            For Each GeneratedName As String In NamesToLookup
                Dim Age As Integer = RandomDriver.Next(PersonLookupMinimumAgeYears, PersonLookupMaximumAgeYears)
                Dim AgeMonth As Integer = RandomDriver.Next(-12, 12)
                Dim AgeDay As Integer = RandomDriver.Next(-31, 31)
                Dim Birthdate As Date = Date.Now.AddYears(-Age).AddMonths(AgeMonth).AddDays(AgeDay)
                Dim FinalAge As Integer = New DateTime((Date.Now - Birthdate).Ticks).Year
                Dim FirstName As String = GeneratedName.Substring(0, GeneratedName.IndexOf(" "))
                Dim LastName As String = GeneratedName.Substring(GeneratedName.IndexOf(" ") + 1)

                'Print all information
                Console.Clear()
                WriteWherePlain("  - Name:                {0}", 0, 1, False, GeneratedName)
                WriteWherePlain("  - First Name:          {0}", 0, 2, False, FirstName)
                WriteWherePlain("  - Last Name / Surname: {0}", 0, 3, False, LastName)
                WriteWherePlain("  - Age:                 {0} years old", 0, 4, False, FinalAge)
                WriteWherePlain("  - Birth date:          {0}", 0, 5, False, Render(Birthdate))

                'Lookup delay
                SleepNoBlock(PersonLookupDelay, ScreensaverDisplayerThread)
            Next

            'Wait until we run the lookup again
            SleepNoBlock(PersonLookupLookedUpDelay, ScreensaverDisplayerThread)
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/PersonLookup.cs
//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Collections.Generic;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using KS.TimeDate;
using Terminaux.Base;
using Textify.NameGen;

namespace KS.Misc.Screensaver.Displays
{
    public static class PersonLookupSettings
    {

        private static int _personLookupDelay = 75;
        private static int _personLookupLookedUpDelay = 10000;
        private static int _personLookupMinimumNames = 10;
        private static int _personLookupMaximumNames = 100;
        private static int _personLookupMinimumAgeYears = 18;
        private static int _personLookupMaximumAgeYears = 100;

        /// <summary>
        /// [PersonLookup] How many milliseconds to wait before getting the new name?
        /// </summary>
        public static int PersonLookupDelay
        {
            get
            {
                return _personLookupDelay;
            }
            set
            {
                if (value <= 0)
                    value = 75;
                _personLookupDelay = value;
            }
        }
        /// <summary>
        /// [PersonLookup] How many milliseconds to show the looked up name?
        /// </summary>
        public static int PersonLookupLookedUpDelay
        {
            get
            {
                return _personLookupLookedUpDelay;
            }
            set
            {
                if (value <= 0)
                    value = 10000;
                _personLookupLookedUpDelay = value;
            }
        }
        /// <summary>
        /// [PersonLookup] Minimum names count
        /// </summary>
        public static int PersonLookupMinimumNames
        {
            get
            {
                return _personLookupMinimumNames;
            }
            set
            {
                if (value <= 10)
                    value = 10;
                if (value > 1000)
                    value = 1000;
                _personLookupMinimumNames = value;
            }
        }
        /// <summary>
        /// [PersonLookup] Maximum names count
        /// </summary>
        public static int PersonLookupMaximumNames
        {
            get
            {
                return _personLookupMaximumNames;
            }
            set
            {
                if (value <= _personLookupMinimumNames)
                    value = _personLookupMinimumNames;
                if (value > 1000)
                    value = 1000;
                _personLookupMaximumNames = value;
            }
        }
        /// <summary>
        /// [PersonLookup] Minimum age years
        /// </summary>
        public static int PersonLookupMinimumAgeYears
        {
            get
            {
                return _personLookupMinimumAgeYears;
            }
            set
            {
                if (value <= 18)
                    value = 18;
                if (value > 100)
                    value = 100;
                _personLookupMinimumAgeYears = value;
            }
        }
        /// <summary>
        /// [PersonLookup] Maximum age years
        /// </summary>
        public static int PersonLookupMaximumAgeYears
        {
            get
            {
                return _personLookupMaximumAgeYears;
            }
            set
            {
                if (value <= _personLookupMinimumAgeYears)
                    value = _personLookupMinimumAgeYears;
                if (value > 100)
                    value = 100;
                _personLookupMaximumAgeYears = value;
            }
        }

    }

    public class PersonLookupDisplay : BaseScreensaver, IScreensaver
    {

        private Random RandomDriver;
        public override string ScreensaverName { get; set; } = "PersonLookup";

        public override Dictionary<string, object> ScreensaverSettings { get; set; }

        public override void ScreensaverPreparation()
        {
            // Variable preparations
            RandomDriver = new Random();
            NameGenerator.PopulateNames();
        }

        public override void ScreensaverLogic()
        {
            Console.BackgroundColor = ConsoleColor.Black;
            Console.ForegroundColor = ConsoleColor.Green;
            ConsoleWrapper.Clear();
            ConsoleWrapper.CursorVisible = false;

            // Generate names
            int NumberOfPeople = RandomDriver.Next(PersonLookupSettings.PersonLookupMinimumNames, PersonLookupSettings.PersonLookupMaximumNames);
            string[] NamesToLookup = NameGenerator.GenerateNames(NumberOfPeople);

            // Loop through names
            foreach (string GeneratedName in NamesToLookup)
            {
                int Age = RandomDriver.Next(PersonLookupSettings.PersonLookupMinimumAgeYears, PersonLookupSettings.PersonLookupMaximumAgeYears);
                int AgeMonth = RandomDriver.Next(-12, 12);
                int AgeDay = RandomDriver.Next(-31, 31);
                var Birthdate = DateTime.Now.AddYears(-Age).AddMonths(AgeMonth).AddDays(AgeDay);
                int FinalAge = new DateTime((DateTime.Now - Birthdate).Ticks).Year;
                string FirstName = GeneratedName.Substring(0, GeneratedName.IndexOf(" "));
                string LastName = GeneratedName.Substring(GeneratedName.IndexOf(" ") + 1);

                // Print all information
                ConsoleWrapper.Clear();
                TextWriterWhereColor.WriteWherePlain("  - Name:                {0}", 0, 1, false, GeneratedName);
                TextWriterWhereColor.WriteWherePlain("  - First Name:          {0}", 0, 2, false, FirstName);
                TextWriterWhereColor.WriteWherePlain("  - Last Name / Surname: {0}", 0, 3, false, LastName);
                TextWriterWhereColor.WriteWherePlain("  - Age:                 {0} years old", 0, 4, false, FinalAge);
                TextWriterWhereColor.WriteWherePlain("  - Birth date:          {0}", 0, 5, false, TimeDateRenderers.Render(Birthdate));

                // Lookup delay
                ThreadManager.SleepNoBlock(PersonLookupSettings.PersonLookupDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
            }

            // Wait until we run the lookup again
            ThreadManager.SleepNoBlock(PersonLookupSettings.PersonLookupLookedUpDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
        }

    }
}